vendor/src/github.com/docker/libnetwork/datastore/datastore.go

```diff
@@ -283,11 +305,31 @@ func (ds *datastore) Watch(kvObject KVObject, stopCh <-chan struct{}) (<-chan KV
 				kvoCh <- dstO
 			}
 		}
+
+		// Wait on watch channel for a re-trigger when datastore becomes active
+		<-watchCh
+
+		kvpCh, err = ds.store.Watch(Key(kvObject.Key()...), sCh)
+		if err != nil {
+			log.Printf("Could not watch the key %s in store: %v", Key(kvObject.Key()...), err)
+		}
+
+		goto retry_watch
```

Comment:
The logic of this Watch wrapper is so intricate and confusing, I had hard times understanding what is was doing and how it was handling the watch failure. This can be simplified by splitting this in two functions, one managing the watch loop and the other handling the retry process (similar to the way it's done in kv discovery in docker for example).

The use of labels is also totally unnecessary and the fact that this `goto` at the end is going back to the beginning of the goroutine screams that you should use a simple loop and/or split the logic in two functions :crying_cat_face: